mgr/dashboard: add create and delete for smb join-auth and usersgroups resources

Pegonzal opened the pull request:
<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

nmunet reviewed (commented):
Hey @Pegonzal,
The work here is looking good, I don't see any major concerns

afreen23 commented:
Hey @Pegonzal lets rebase and take this PR now, thanks :)

afreen23 reviewed (changes requested):
Hi @Pegonzal thanks for the PR 👍🏿 

Just some UX suggestions:

1. Since most of the fields in `join-auth` are `required` lets have only `optional` ones marked as per carbon guidelines.
2. We need to use password input here, there is `cds-passwor-label` which comes with hide/show for password inbuilt . Check [here](https://angular.carbondesignsystem.com/?path=/docs/components-input-password--docs)
3. What is `join-auth`? We need a one-line helper text here cc @bhkaur25 @rivkap . Also is there a better term/generic to use here ?
4. `Linked to Cluster` can be just `Cluster` since we are explaining that in the helper text. Also lets choose a default value there or first item in list if no defaults.
5. We can have some predefined join-auth name here for faster form inputs. User may edit them for sure.

afreen23 commented:
For the usr/group form:
![image](https://github.com/user-attachments/assets/8e76ac5b-b92d-4741-bde6-2397ce6b9177)
1. Delete icon seems off the password field
2. This component is I see reused many places, so lets have a generic component created cc @dnyanee1997 

![image](https://github.com/user-attachments/assets/ee028c21-94bf-4d19-ba2f-051415a2af83)
1. I understand in backend teh resource name is `usersgroups` but we dont need to copy that in UI. It makes sense to call it `Create Users and Groups` alongwith helper text.
2. Instead of `userg group ID` use `Unique Identifier` and helper text can be improved to `A unique name for the `usersandgroups` resource`. And we can have this pre-populated too.
3. Linked to cluster can be `Cluster` ,same as above
4. password fields can be used here, same as above
5. We need to add helper text for `groups`

Thanks

bhkaur25 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/smb/smb-cluster-form/smb-cluster-form.component.html:
```diff
@@ -269,6 +267,43 @@
+             [columnNumbers]="{ lg: 14 }">
+          <input cdsText
+                 [formControlName]="i"
+                 placeholder="Custom DNS"/>
```
Hi @afreen23  @Pegonzal  
Can we name it **smb_user_authentication**
If it’s about authenticating users, this makes the scope clear.

afreen23 replied:
This is actually something like "User authentication" but @Pegonzal can confirm

Pegonzal replied:
> What is join-auth? 

Logical management unit for authentication against active directory samba servers

> Also is there a better term/generic to use here ?

I'm going with Active Directory (AD) access resource for most places or something similar

@afreen

Pegonzal commented:
Hey @afreen23, I've updated the PR. Please take another look.

Most of the stuff I've addressed, for some other things I've opened a tracker to take them after this PR as I found them to be nonblockers and I'd prefer to open one last PR regarding UX enhancements where I can take this. For this I'm also waiting on the content's team feedback to improve helperTexts and overall wording for the terminology, this is the tracker I'll work on next: https://tracker.ceph.com/issues/69993

cloudbehl commented:
Tested it, works well.

Few improvements.

1. small typo: resourcen
![Screenshot from 2025-02-18 15-54-08](https://github.com/user-attachments/assets/a8c28b4c-02eb-440b-ac7e-4d38a3950c6f)
2. Error talks about Join sources and domain setting, now we don't any resource with that name.
![Screenshot from 2025-02-18 16-08-49](https://github.com/user-attachments/assets/70bd4e55-d17c-453d-99ad-56e88f17eee2)
3. While creating the cluster if the AD/Standalone is already linked to another cluster, we are still showing them to get added. Also, I'm able to select the same resource(AD/standalone) multiple time while creating cluster.
![image](https://github.com/user-attachments/assets/d81264dc-4dd5-44e3-88dc-66bcfea5ced8)

Pegonzal commented:
Thank you for testing @cloudbehl 

> 1. small typo: resourcen

Fixed it

> 2. Error talks about Join sources and domain setting, now we don't any resource with that name.

Updated to `Specify the Realm and AD access resources in the Domain Settings field.`

> 3. While creating the cluster if the AD/Standalone is already linked to another cluster, we are still showing them to get added. Also, I'm able to select the same resource(AD/standalone) multiple time while creating cluster.

I've added to this tracker which I'm working on next, but first I would like to confirm there is no reason for the cluster to allow for same id access resources when creating it

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/app-routing.module.ts:
```diff
@@ -457,6 +459,26 @@ const routes: Routes = [
                 data: { breadcrumbs: ActionLabels.CREATE }
+              },
+              {
+                path: `joinauth/${URLVerbs.CREATE}`,
+                component: SmbJoinAuthFormComponent,
+                data: { breadcrumbs: ActionLabels.CREATE }
+              },
+              {
+                path: `joinauth/${URLVerbs.EDIT}/:authId`,
+                component: SmbJoinAuthFormComponent,
+                data: { breadcrumbs: ActionLabels.EDIT }
+              },
+              {
+                path: `usersgroups/${URLVerbs.CREATE}`,
+                component: SmbUsersgroupsFormComponent,
+                data: { breadcrumbs: ActionLabels.CREATE }
+              },
+              {
+                path: `usersgroups/${URLVerbs.EDIT}/:usersGroupsId`,
+                component: SmbUsersgroupsFormComponent,
+                data: { breadcrumbs: ActionLabels.EDIT }
```
urls should match with what we are mentioning in UI

Pegonzal replied:
Changed it to ad and standalone as create buttons are displaying

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/smb/smb-cluster-form/smb-cluster-form.component.ts:
```diff
@@ -43,13 +45,15 @@ export class SmbClusterFormComponent extends CdForm implements OnInit {
   hasOrchestrator: boolean;
   orchStatus$: Observable<any>;
   allClustering: string[] = [];
+  CLUSTERING = CLUSTERING;
```
it would be nice to organize all constants together with Icons enum

Pegonzal replied:
I'm sorry I do not understand this

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/smb/smb-cluster-form/smb-cluster-form.component.ts:
```diff
@@ -241,6 +247,22 @@ export class SmbClusterFormComponent extends CdForm implements OnInit {
 
+    if (rawFormValue.public_addrs?.length > 0) {
+      requestModel.cluster_resource.public_addrs = [];
+      for (let i = 0; i < rawFormValue.public_addrs.length; i++) {
+        if (rawFormValue.public_addrs[i].destination) {
+          requestModel.cluster_resource.public_addrs.push({
+            address: rawFormValue.public_addrs[i].address,
+            destination: rawFormValue.public_addrs[i].destination
+          });
```
what about using map/filter ?

Pegonzal replied:
Using map now

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/smb/smb-join-auth-form/smb-join-auth-form.component.html:
```diff
@@ -0,0 +1,126 @@
+          id="authId"
+          name="authId"
+          formControlName="authId"
+          [invalid]="form.controls.authId.invalid && form.controls.authId.dirty"
```
I think `showError` checks for both ? may be worth checking if
Similarly for others below

Pegonzal replied:
With dynamic forms I've had problems with this due to the `this.get(controlName)` not finding the controlname in the form.

I've tried to implement showError instead of current form.control validator but it only accepts `NgForm` as paramter for the form, my form here is a `CdFormGroup`, did not look much else into it, so leaving it as  it is for now

afreen23 reviewed (changes requested):
Unit tests are missing, we should add those.
Thanks

Pegonzal commented:
> Unit tests are missing, we should add those. Thanks

@afreen23  I've added some unit tests for joinauth and usersgroups resources both in front and backend, as well as create form

Pegonzal commented:
jenkins test api

Pegonzal commented:
jenkins test make check

Pegonzal commented:
jenkins retest this please

Pegonzal commented:
Jenkins test api
